src/rgw/rgw_perf_counters.cc

```diff
@@ -3,69 +3,207 @@
+  add_rgw_op_counters(&pcb);
+  PerfCounters *new_counters = pcb.create_perf_counters();
+  cct->get_perfcounters_collection()->add(new_counters);
+  global_op_counters = new_counters;
+}
+
+CountersContainer get(req_state *s) {
+  CountersContainer counters;
+
+  if (user_counters_cache && !s->user->get_id().id.empty()) {
+    std::string key = ceph::perf_counters::key_create(rgw_op_counters_key, {{"User", s->user->get_id().id}});
+    counters.user_counters = user_counters_cache->get(key);
+  }
+
+  if (bucket_counters_cache && !s->bucket_name.empty()) {
+    std::string key = ceph::perf_counters::key_create(rgw_op_counters_key, {{"Bucket", s->bucket_name}});
+    counters.bucket_counters = bucket_counters_cache->get(key);
+  }
+
+  return counters;
+}
+
+void inc(const CountersContainer &counters, int idx, uint64_t v) {
+  if (counters.user_counters) {
+    PerfCounters *user_counters = counters.user_counters.get();
+    user_counters->inc(idx, v);
+  }
+  if (counters.bucket_counters) {
+    PerfCounters *bucket_counters = counters.bucket_counters.get();
+    bucket_counters->inc(idx, v);
+  }
+  if (global_op_counters) {
+    global_op_counters->inc(idx, v);
+  }
+}
+
+void tinc(const CountersContainer &counters, int idx, utime_t amt) {
+  if (counters.user_counters) {
+    PerfCounters *user_counters = counters.user_counters.get();
+    user_counters->tinc(idx, amt);
+  }
+  if (counters.bucket_counters) {
+    PerfCounters *bucket_counters = counters.bucket_counters.get();
+    bucket_counters->tinc(idx, amt);
+  }
+  if (global_op_counters) {
+    global_op_counters->tinc(idx, amt);
+  }
+}
+
+void tinc(const CountersContainer &counters, int idx, ceph::timespan amt) {
+  if (counters.user_counters) {
+    PerfCounters *user_counters = counters.user_counters.get();
+    user_counters->tinc(idx, amt);
+  }
+  if (counters.bucket_counters) {
+    PerfCounters *bucket_counters = counters.bucket_counters.get();
+    bucket_counters->tinc(idx, amt);
+  }
+  if (global_op_counters) {
+    global_op_counters->tinc(idx, amt);
+  }
+}
+
+} // namespace rgw::op_counters
+
+int rgw_perf_start(CephContext *cct)
+{
+  frontend_counters_init(cct);
+
+  bool user_counters_cache_enabled = cct->_conf.get_val<bool>("rgw_user_counters_cache");
+  if (user_counters_cache_enabled) {
+    uint64_t target_size = cct->_conf.get_val<uint64_t>("rgw_user_counters_cache_size");
+    rgw::op_counters::user_counters_cache = new ceph::perf_counters::PerfCountersCache(cct, target_size, create_rgw_counters);
+  }
+
+  bool bucket_counters_cache_enabled = cct->_conf.get_val<bool>("rgw_bucket_counters_cache");
+  if (bucket_counters_cache_enabled) {
+    uint64_t target_size = cct->_conf.get_val<uint64_t>("rgw_bucket_counters_cache_size");
+    rgw::op_counters::bucket_counters_cache = new ceph::perf_counters::PerfCountersCache(cct, target_size, create_rgw_counters);
+  }
+
+  rgw::op_counters::global_op_counters_init(cct);
```

Comment:
nit: namespacing is a good way to avoid duplicating names. consider `rgw::op_counters::global_init()` instead

similar feedback for `ceph::perf_counters::PerfCountersCache`, `rgw::op_counters::global_op_counters`, `rgw::op_counters::user_counters_cache` etc